File: macOS/App/TigerTVRepository.swift
import Foundation

actor TigerTVRepository {
    private let configDataSource: ConfigDataSource
    private let apiClient: MacCMSApiClient
    private let playbackResolver: PlaybackURLResolver
    private let searchCacheTtl: TimeInterval

    private var cachedConfig: SourceConfig?
    private var searchCache: [String: CacheEntry<SearchResponse>] = [:]
    private var coverCache: [String: URL] = [:]
    private let coverSemaphore = AsyncSemaphore(value: 4)
    /// 兜底 URL 的磁盘持久层（7 天 TTL）；nil 表示仅进程内缓存。
    private let coverURLStore: CoverFallbackURLStore?

    init(
        configDataSource: ConfigDataSource,
        apiClient: MacCMSApiClient,
        playbackResolver: PlaybackURLResolver,
        searchCacheTtl: TimeInterval = 10 * 60,
        coverURLStore: CoverFallbackURLStore? = nil
    ) {
        self.configDataSource = configDataSource
        self.apiClient = apiClient
        self.playbackResolver = playbackResolver
        self.searchCacheTtl = searchCacheTtl
        self.coverURLStore = coverURLStore
    }

    func loadConfig() async -> LoadResult<SourceConfig> {
        let result = await configDataSource.loadConfig()
        if case .success(let config) = result {
            cachedConfig = config
        }
        return result
    }

    func search(keyword: String) async -> LoadResult<SearchResponse> {
        let trimmed = keyword.trimmingCharacters(in: .whitespacesAndNewlines)

        if let cached = searchCache[trimmed], cached.isFresh(ttl: searchCacheTtl) {
            return .success(cached.data)
        }

        let config: SourceConfig
        switch await loadConfig() {
        case .success(let c): config = c
        case .failure(let e): return .failure(e)
        }

        let sites = Array(config.apiSite.values)
        let semaphore = AsyncSemaphore(value: 20)
        let results = await withTaskGroup(of: [SearchResult].self) { group in
            for site in sites {
                group.addTask {
                    await self.searchOne(site: site, keyword: trimmed, semaphore: semaphore)
                }
            }
            var collected: [SearchResult] = []
            for await partial in group {
                collected.append(contentsOf: partial)
            }
            return collected
        }

        let response = SearchResponse(keyword: trimmed, results: results)
        searchCache[trimmed] = CacheEntry(data: response, createdAt: Date())
        return .success(response)
    }

    func fetch(siteName: String, vodId: Int) async -> LoadResult<FetchResponse> {
        let config: SourceConfig
        switch await loadConfig() {
        case .success(let c): config = c
        case .failure(let e): return .failure(e)
        }

        guard let site = config.apiSite.values.first(where: { $0.name == siteName }) else {
            let available = config.apiSite.values.map(\.name).sorted()
            return .failure(TigerTVError.unknownSite(siteName, available))
        }

        do {
            let detail = try await apiClient.fetchDetail(api: site.api, vodId: vodId)
            guard detail.code == 1 else {
                return .failure(TigerTVError.fetchFailed(detail.msg ?? "code \(detail.code)"))
            }
            guard let item = detail.list.first else {
                return .failure(TigerTVError.fetchFailed("空详情响应"))
            }
            let play = parseEpisodeLinks(raw: item.vodPlayUrl ?? "")
            let down = parseEpisodeLinks(raw: item.vodDownUrl ?? "")
            let response = FetchResponse(
                vodId: item.vodId,
                site: siteName,
                vodPlayUrl: play,
                vodDownUrl: down
            )
            return .success(response)
        } catch {
            return .failure(TigerTVError.network(error.localizedDescription))
        }
    }

    func resolvePlaybackUrl(url: String) async -> LoadResult<String> {
        do {
            let resolved = try await playbackResolver.resolve(url)
            return .success(resolved)
        } catch let error as TigerTVError {
            return .failure(error)
        } catch {
            return .failure(.network(error.localizedDescription))
        }
    }

    /// 搜索封面懒加载兜底：多数站点 `ac=list` 响应不含 `vod_pic`（仅 detail 返回），
    /// 空封面卡片可见时按需取 detail 补齐。命中内存缓存不发请求；失败不缓存，
    /// 卡片再次进入视口时自然重试。并发受信号量限制，避免滚动时打爆站点。
    func coverFallbackURL(siteName: String, vodId: Int) async -> URL? {
        let key = "\(siteName)-\(vodId)"
        if let cached = coverCache[key] { return cached }

        // 磁盘持久层命中：回填内存缓存即可，无需发 detail 请求（7 天 TTL 内稳定有效）。
        if let persisted = coverURLStore?.url(forKey: key),
           let url = HTTPClient.percentEncodedURL(from: persisted) {
            coverCache[key] = url
            return url
        }

        let config: SourceConfig
        switch await loadConfig() {
        case .success(let c): config = c
        case .failure: return nil
        }
        guard let site = config.apiSite.values.first(where: { $0.name == siteName }) else {
            return nil
        }

        await coverSemaphore.wait()
        // 拿到许可后二次检查：排队期间同 key 可能已被其他任务填入缓存。
        if let cached = coverCache[key] {
            await coverSemaphore.signal()
            return cached
        }
        do {
            let detail = try await apiClient.fetchDetail(api: site.api, vodId: vodId)
            await coverSemaphore.signal()
            guard detail.code == 1,
                  let pic = detail.list.first?.vodPic?.trimmingCharacters(in: .whitespacesAndNewlines),
                  !pic.isEmpty else {
                return nil
            }
            let url = HTTPClient.percentEncodedURL(from: pic)
            if let url {
                coverCache[key] = url
                coverURLStore?.store(pic, forKey: key)
            }
            return url
        } catch {
            await coverSemaphore.signal()
            return nil
        }
    }

    private func searchOne(site: SourceSite, keyword: String, semaphore: AsyncSemaphore) async -> [SearchResult] {
        await semaphore.wait()
        let results = await apiClient.search(siteName: site.name, api: site.api, keyword: keyword)
        await semaphore.signal()
        return results
    }

    private func parseEpisodeLinks(raw: String) -> [EpisodeLink] {
        guard !raw.isEmpty else { return [] }
        var links: [EpisodeLink] = []
        let groups = raw.components(separatedBy: "$$$")
        for group in groups {
            let items = group.components(separatedBy: "#")
            for item in items {
                let parts = item.split(separator: "$", maxSplits: 1, omittingEmptySubsequences: false)
                guard parts.count == 2 else { continue }
                let name = String(parts[0]).trimmingCharacters(in: .whitespaces)
                let url = String(parts[1]).trimmingCharacters(in: .whitespaces)
                if !name.isEmpty && !url.isEmpty {
                    links.append(EpisodeLink(name: name, url: url))
                }
            }
        }
        return links
    }
}

private struct CacheEntry<T> {
    let data: T
    let createdAt: Date

    func isFresh(ttl: TimeInterval) -> Bool {
        Date().timeIntervalSince(createdAt) < ttl
    }
}

private actor AsyncSemaphore {
    private var value: Int
    private var waiters: [CheckedContinuation<Void, Never>] = []

    init(value: Int) {
        self.value = value
    }

    func wait() async {
        if value > 0 {
            value -= 1
            return
        }
        await withCheckedContinuation { continuation in
            waiters.append(continuation)
        }
    }

    func signal() {
        if let continuation = waiters.first {
            waiters.removeFirst()
            continuation.resume()
        } else {
            value += 1
        }
    }
}

File: AndroidTV/app/src/main/java/art/anjing/tigertv/data/TigerTVRepository.kt
package art.anjing.tigertv.data

import android.util.Log
import art.anjing.tigertv.domain.EpisodeLink
import art.anjing.tigertv.domain.FetchResponse
import art.anjing.tigertv.domain.SearchResponse
import art.anjing.tigertv.domain.SearchResult
import art.anjing.tigertv.domain.SourceConfig
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.sync.withPermit
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File
import java.util.concurrent.CancellationException

class TigerTVRepository(
    private val configDataSource: ConfigDataSource,
    private val apiClient: MacCMSApiClient,
    private val playbackUrlResolver: PlaybackUrlResolver,
    private val searchCacheTtlMillis: Long = DEFAULT_SEARCH_CACHE_TTL,
    coverUrlCacheDir: File? = null
) {
    private var cachedConfig: SourceConfig? = null
    private val searchCache = mutableMapOf<String, CacheEntry<SearchResponse>>()
    private val coverCache = mutableMapOf<String, String>()
    private val coverCacheMutex = Mutex()
    private val coverSemaphore = Semaphore(COVER_FALLBACK_CONCURRENCY)

    /** 兜底 URL 磁盘持久层（7 天 TTL）；null 表示仅进程内缓存。 */
    private val coverDiskCache = coverUrlCacheDir?.let { CoverFallbackDiskCache(it) }

    /**
     * 搜索封面懒加载兜底：多数站点 `ac=list` 响应不含 `vod_pic`（仅 detail 返回），
     * 空封面卡片可见时按需取 detail 补齐。命中内存缓存不发请求；失败不缓存，
     * 卡片再次进入视口时自然重试。并发受信号量限制，避免滚动时打爆站点。
     */
    suspend fun resolveCoverFallback(siteName: String, vodId: Int): String? {
        val key = "$siteName-$vodId"
        coverCacheMutex.withLock {
            coverCache[key]?.let { return it }
        }
        // 磁盘持久层命中：回填内存缓存即可，无需发 detail 请求（7 天 TTL 内稳定有效）。
        coverDiskCache?.get(key)?.let { persisted ->
            coverCacheMutex.withLock { coverCache[key] = persisted }
            return persisted
        }

        val config = cachedConfig ?: when (val loaded = loadConfig()) {
            is Result.Success -> loaded.data
            is Result.Error -> return null
        }
        val site = config.apiSite.values.find { it.name == siteName } ?: return null

        return try {
            coverSemaphore.withPermit {
                // 拿到许可后二次检查：排队期间同 key 可能已被其他任务填入缓存。
                val cached = coverCacheMutex.withLock { coverCache["$siteName-$vodId"] }
                cached ?: run {
                    val detail = apiClient.fetchDetail(site.api, vodId)
                    val pic = if (detail.code == 1) detail.list.firstOrNull()?.vodPic?.trim() else null
                    val url = pic?.takeIf { it.isNotEmpty() }
                    if (url != null) {
                        coverCacheMutex.withLock { coverCache["$siteName-$vodId"] = url }
                        coverDiskCache?.put("$siteName-$vodId", url)
                    }
                    url
                }
            }
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            Log.w(TAG, "Cover fallback failed for $siteName/$vodId", e)
            null
        }
    }

    suspend fun loadConfig(): Result<SourceConfig> {
        return when (val result = configDataSource.loadConfig()) {
            is Result.Success -> {
                cachedConfig = result.data
                result
            }
            is Result.Error -> result
        }
    }

    suspend fun search(keyword: String): Result<SearchResponse> {
        val trimmed = keyword.trim()
        val cached = searchCache[trimmed]
        if (cached != null && cached.isFresh(searchCacheTtlMillis)) {
            Log.i(TAG, "Using cached search results for '$trimmed'")
            return Result.Success(cached.data)
        }

        val config = cachedConfig ?: when (val loaded = loadConfig()) {
            is Result.Success -> loaded.data
            is Result.Error -> return Result.Error(loaded.exception)
        }

        return try {
            val sites = config.apiSite.values.toList()
            val semaphore = Semaphore(MAX_CONCURRENT)
            val results = coroutineScope {
                sites.map { site ->
                    async {
                        semaphore.withPermit {
                            apiClient.search(site.name, site.api, trimmed)
                        }
                    }
                }.awaitAll().flatten()
            }
            val response = SearchResponse(keyword, results)
            searchCache[trimmed] = CacheEntry(response, System.currentTimeMillis())
            Result.Success(response)
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            Result.Error(e)
        }
    }

    suspend fun fetch(siteName: String, vodId: Int): Result<FetchResponse> {
        val config = cachedConfig ?: when (val loaded = loadConfig()) {
            is Result.Success -> loaded.data
            is Result.Error -> return Result.Error(loaded.exception)
        }

        val site = config.apiSite.values.find { it.name == siteName }
            ?: return Result.Error(IllegalArgumentException("Unknown site: $siteName. Available: ${config.apiSite.values.map { it.name }}"))

        return try {
            val detail = apiClient.fetchDetail(site.api, vodId)
            if (detail.code != 1) {
                return Result.Error(IllegalStateException("Fetch failed: ${detail.msg}"))
            }
            val item = detail.list.firstOrNull()
                ?: return Result.Error(IllegalStateException("Empty detail response"))
            val play = parseEpisodeLinks(item.vodPlayUrl)
            val down = parseEpisodeLinks(item.vodDownUrl)
            Result.Success(
                FetchResponse(
                    vodId = item.vodId,
                    site = siteName,
                    vodPlayUrl = play,
                    vodDownUrl = down
                )
            )
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            Result.Error(e)
        }
    }

    suspend fun resolvePlaybackUrl(url: String): Result<String> {
        return try {
            Result.Success(playbackUrlResolver.resolve(url))
        } catch (e: CancellationException) {
            // 结构化取消必须重抛，否则 viewModelScope.cancel 无法中断子 job。
            throw e
        } catch (e: Exception) {
            Result.Error(e)
        }
    }

    private fun parseEpisodeLinks(raw: String): List<EpisodeLink> {
        if (raw.isBlank()) return emptyList()
        val links = mutableListOf<EpisodeLink>()
        val groups = raw.split("$$$")
        for (group in groups) {
            val items = group.split("#")
            for (item in items) {
                val parts = item.split("$", limit = 2)
                val name = parts.getOrNull(0)?.trim() ?: continue
                val url = parts.getOrNull(1)?.trim() ?: continue
                if (name.isNotEmpty() && url.isNotEmpty()) {
                    links.add(EpisodeLink(name, url))
                }
            }
        }
        return links
    }

    companion object {
        private const val MAX_CONCURRENT = 20
        private const val COVER_FALLBACK_CONCURRENCY = 4
        private const val DEFAULT_SEARCH_CACHE_TTL = 10 * 60 * 1000L // 10 minutes
        private const val TAG = "TigerTVRepository"
    }

    private data class CacheEntry<T>(
        val data: T,
        val createdAtMillis: Long
    ) {
        fun isFresh(ttlMillis: Long): Boolean {
            return System.currentTimeMillis() - createdAtMillis < ttlMillis
        }
    }
}

/**
 * 兜底封面 URL 的磁盘持久缓存（JSON + 原子替换，7 天 TTL）。
 * 进程重启后命中即可跳过 detail 请求；读取时惰性清除过期项。
 */
internal class CoverFallbackDiskCache(private val directory: File) {
    @Serializable
    data class Entry(val url: String, val fetchedAtMillis: Long)

    private val file = File(directory, "cover-fallback-cache.json")
    private val mutex = Mutex()
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun get(key: String): String? = mutex.withLock {
        val entries = read()
        val entry = entries[key] ?: return@withLock null
        if (System.currentTimeMillis() - entry.fetchedAtMillis >= TTL_MILLIS) {
            write(entries - key)
            null
        } else {
            entry.url
        }
    }

    suspend fun put(key: String, url: String) {
        put(key, url, System.currentTimeMillis())
    }

    internal suspend fun put(key: String, url: String, fetchedAtMillis: Long) = mutex.withLock {
        write(read() + (key to Entry(url, fetchedAtMillis)))
    }

    private fun read(): Map<String, Entry> = try {
        json.decodeFromString<Map<String, Entry>>(file.readText())
    } catch (_: Exception) {
        emptyMap()
    }

    private fun write(entries: Map<String, Entry>) {
        try {
            directory.mkdirs()
            // tmp + rename 原子写，避免进程中断留下半截 JSON。
            val tmp = File(directory, "${file.name}.tmp")
            tmp.writeText(json.encodeToString(entries))
            if (!tmp.renameTo(file)) {
                file.delete()
                tmp.renameTo(file)
            }
        } catch (_: Exception) {
            // 缓存写失败不影响主流程。
        }
    }

    companion object {
        const val TTL_MILLIS = 7L * 24 * 60 * 60 * 1000
    }
}
